### ISSUE TITLE
new model paper/details

### ISSUE THREAD

**Issue by G-Wang** [fatchord/WaveRNN#9]
Hi, you new model sounds very good, any chance you will write it up in a paper/blog-post? What's the new vocoder, it's it more WaveRNN like or wavenet like?

**Comment by fatchord** [comment]
@G-Wang Thanks for the kind words. I'm not in academia so writing a paper isn't really on the cards - I guess a picture paints a thousand words hence the diagram. Anyway, hope you find the model useful - I've personally found it much faster and easier to train than wavenet. 

The output is a little noisy though - but that probably could be improved with some more experimentation. Perhaps using a Gaussian (like Baidu's Clarinet paper) would work better - let's see.

**Comment by pfriesch** [comment]
In your alternative model, it seems like you don't use the dual softmax anymore. Did you experiment with it?

**Comment by G-Wang** [comment]
@fatchord Absolutely, your model converges so much faster and easier to train than all the wavenet implementations, I can get pretty fast training even on a measly gtx 1060.

I was always a bit thrown off by the earliest wavenet paper when they found that softmax logits are easier to train than Gaussian.

Have you done anything similar to WaveRNN paper's pruning of their networks? Althogh this implementation is already easy on GPU memory and fast for inference

**Comment by fatchord** [comment]
@G-Wang re:ablation studies - that's a good idea so I'll reopen this. I've wondered whether the resnet is really necessary, so it would be nice to retrain it without it. Ditto with the last dense layer too. 

Another thing I should mention - I tried a 10bit and 11bit signal and neither sound as good as 9bit at ~500k steps of training. It could simply be a case of needing more parameters or training steps.

The one good thing about the resnet is that it can be run in parallel - that's one of the main reasons I put it in in the first place. Perhaps deepening it significantly would help too.

The only problem is that my GPU is tied up at the minute so I can't run the experiments right now.

**Comment by fatchord** [comment]
@pfriesch I did and wasn't too happy with the results thus far. That's not to say that it's impossible, just had no luck with it using CUDA rnn layers.

**Comment by G-Wang** [comment]
@fatchord I have some compute so I'll run some higher bit runs to see whether further training/tuning will help

**Comment by fatchord** [comment]
@G-Wang Thanks, that would be really appreciated!

**Comment by jjj8080** [comment]
I have spare compute too, how many blocks deep do you think could be ideal with the ResNet?

Great work by the way!

**Comment by G-Wang** [comment]
@fatchord Given that the alternative model predicts over the bits directly (without splitting it up into course/fine), I'm not sure whether a softmax over 4096 classes (for 12 bit audio) can perform that well. While I'm training the higher bit-rates I'll see if I can get Clarinet's Gaussian output prediction working (via MLE).

**Comment by fatchord** [comment]
@G-Wang I agree - going up to 12bit is probably a little pointless without a big increase in parameters/layers - which goes against the purpose of WaveRNN in the first place. I'd be much more interested in either removing the resnet or trying a Gaussian. No pressure though, only do it if you have the time & compute to spare - I don't want to impose.

**Comment by MlWoo** [comment]
@fatchord  I am impressed with your work.  There are 2 questions:

1.  What is your dataset?
2.  How much is the loss the model achieved when training ?

Thank you in advance.

**Comment by fatchord** [comment]
@MlWoo Thanks, the data is from  "Richard Nixon: The Life by John A. Farrell" audiobook. The loss is my last training run got to around 1.

**Comment by MlWoo** [comment]
@fatchord That's awsome! The model converges very quickly, howerver, my model got 1.5 at around 90 epochs. How many epochs did you model cost to get to around 1.

**Comment by fatchord** [comment]
@MlWoo I normally train for around 500k steps at a batch size of 16. I also notice that, while it may seem like the model has converged much earlier than 500k, you need to keep training it. The sound quality does improve even though it looks like it's stuck in a saddle point.

Not sure why there's such a discrepancy in our respective loss value, perhaps you are using different hyperparameters? Also I'd love to hear samples of the model you trained - feel free to post them here!

**Comment by MlWoo** [comment]
@fatchord  Thanks a lot, I will keep training the model. BTW and post them later. Could I konw the tools to sgement the audio into mutiple small audios?

**Comment by geneing** [comment]
@fatchord 
> Another thing I should mention - I tried a 10bit and 11bit signal and neither sound as good as 9bit at ~500k steps of training. It could simply be a case of needing more parameters or training steps.
> 
I trained with 10bit output and I find that the sound quality after ~700k steps is better than for the 9bit output. In particular, 9bit output sounded a little bit noisier (low level white noise in voiced parts). Here's the output for my fork: https://app.box.com/s/wy5wai8tutjir07e66xps0a38mxbdqmy

**Comment by G-Wang** [comment]
@geneing did you play around with seq_len to see what effects it has? For my experience I found that it degraded model performance. However training with seq_len 5 and batch size of 16 really doesn't maximize the GPU usage.

**Comment by G-Wang** [comment]
@fatchord I also made a seperate repo to refractor the code as well as add in a few things. https://github.com/G-Wang/WaveRNN-Pytorch

I made attempts at training a single beta distribution (similar to the clarinet paper), as well as the mixture of logistics from r9r9's Wavenet (this is not that great yet).

Also added in a batch synthesis (still experimental) where during generation we split the mel spectrogram, batch up the split mel spectrograms, generate batch of audio, and stitch the audio back into a single wav. This allows me to get close to real time synthesis, and from the experiments I've run there's not much audio degradation by batching this way. This way we can fully utilize the GPU for synthesis.

**Comment by geneing** [comment]
@G-Wang 
> @geneing did you play around with seq_len to see what effects it has? 
I tried different training lengths for wavenet vocoder and found the same thing. By default everyone trains on 8000 raw audio datapoints. I tried increasing it to 22050 (1 sec). First, loss function was decreasing slower. Second, on test data vocoder was producing phasing(?) errors in sibilant sounds (basically "s" sounded like a short burst of white noise), which were not present when trained on shorter fragments. I think this happens because the network learns to model short term dependencies (for wavenet the receptive field is 1024 if I remember correctly).

Splitting mel spectrograms and generating in parallel is an excellent idea! I wonder if one can improve on it by splitting at well chosen points where phase mismatch and previous history wouldn't matter, such as unvoiced parts or fricatives.

**Comment by G-Wang** [comment]
Yes, I was thinking of splitting with overlaps in Mel , so so we give the model room to generate the right wav.

**Comment by fatchord** [comment]
@geneing 
> I trained with 10bit output and I find that the sound quality after ~700k steps is better than for the 9bit output. In particular, 9bit output sounded a little bit noisier (low level white noise in voiced parts). Here's the output for my fork: https://app.box.com/s/wy5wai8tutjir07e66xps0a38mxbdqmy

They sound really good - there's definitely a reduction in noise compared to 9bit.

**Comment by fatchord** [comment]
@G-Wang 

> Also added in a batch synthesis (still experimental) where during generation we split the mel spectrogram, batch up the split mel spectrograms, generate batch of audio, and stitch the audio back into a single wav.

This is a great idea! Have you any samples from the experiment you can share?

**Comment by G-Wang** [comment]
@fatchord Here're some samples for you.

I've included samples for batch synthesis, where the speed is fast than real time (around 2 seconds for a 6 - 9 second audio). The single synthesis is if we generate the entire mel spectrogram with the model, this takes around 1.5 minutes.

These are all done on a GTX 1060 ti.

As you can hear, there're little clicking artifacts in the batched synthesis version where the batched wav are joined, I created a version where I'm generating overlaps, and then i average the overlaped wav values, but I'm not able to remove the clicking artifacts, any idea for removing these?

**Comment by G-Wang** [comment]
@fatchord I see, I think the low energy/fricatives method is worth investigating, in this case I guess I would have to write a function to look at the melspectrogram conditioning to look for low-energy areas/fricatives where I could split?

I have had some helpful ideas on some post-processing methods as well and I'll investigate those.

**Comment by fatchord** [comment]
@G-Wang I'm guessing that using a simple threshold or combination of thresholds would find appropriate split points. Perhaps splitting into two bands - one looking for low energy in all bands and another looking for lots of energy in the high bands with absence of energy lower down (searching for bright white-ish noise which isn't as phase dependant). That's just off the top of my head - I haven't really looked into it much as of yet.

**Comment by geneing** [comment]
I believe there is another problem with batch generation. The current implementation generates wave files that is a little short compared to the mel data. 

Wave data should be exactly hop_size*num_mels, but because of padding and trimming it's a bit shorter. In fact the sound is speed up by a little bit to fit in slightly shorter sound file. It's not noticeable for a full sentence, but when batching data the speech becomes faster and faster with higher batch sizes. I fixed this problem in my fork. 

The second problem is that padding used in the current code is probably not enough. When I look at the beginning of the sound wave for each fragment, I see that it increases in volume. I had to pad a little more, otherwise sound volume would fluctuate with the number of batches used for generation.

**Comment by G-Wang** [comment]
@geneing yes, in my current implementation if we batch the mel too much the upsampled wav speed is increased alot as you said.
What I ended up doing in my repo (I didn't update yet) was to batch after the mel have been upsampled by the upsampling network.

However I'm still getting these clicking sound in my audio. Do you have this issue with extra padding?

**Comment by fatchord** [comment]
@G-Wang I had a go at batched inference. Link to samples:

[wavernn_batched_inference.tar.gz](https://github.com/fatchord/WaveRNN/files/2649889/wavernn_batched_inference.tar.gz)

I got rid of most of the clicks and pops by upping the float precision to double when doing the crossfading operations (in case any quantisation error caused nyquist reflections). I also gave the rnn half the crossfade length in samples to 'shake off' its early, volatile state (that's when most clicks occur in my experience).

**Comment by G-Wang** [comment]
@fatchord @geneing sounds really good! will you be releasing the implementation by any chance?

Also what's the batch size and generation speed you get with this implementation?

**Comment by geneing** [comment]
@G-Wang Right now I'm working on this branch: https://github.com/geneing/WaveRNN-Pytorch/tree/model_simplification. It is forked from your repo. I like your port better than my own, so I migrated my changes here.

**Comment by fatchord** [comment]
@geneing 

> How are you doing the crossfading?

It's actually pretty simple - like how a dj would mix two records if they used numpy instead of a mixer!

Basically it's `y1 * sqrt(1/2 * (1 - t)) + y2 * sqrt(1/2 * (1 + t)) for t in range -1 to 1`. I was genuinely shocked that this method produces almost no phase issues (as long as the crossfade is nice and tight). 

@G-Wang I'll upload the implementation soon. I'm in the tricky situation where I'm not happy with my new model but quite happy with the batched generation I coded on top of it. I guess I could just retrofit the model in notebook 4b with the batched inference and upload that instead?

As for speed. I let the batchsize be variable and dictated by the tuneable parameters: the target amount of samples for each batch entry and  the overlap (for rnn warm up and crossfade). I'm getting ~20kHz for short samples and 100kHz+ for longer samples. Theoretically, if the frontend feeding the model queued the mels in an efficient manner then I see no reason why something like 200kHz couldn't be achieved - I'm not even close to maxing out my gpu mem/cores with the current implementation.

One thing though - I do get pops every now and again - I'd say roughly one in twenty samples produced have some artifact so it's not 100% yet.

**Comment by fatchord** [comment]
@G-Wang @geneing I updated notebook4b with batched inference if you wanna check it out.

**Comment by geneing** [comment]
@fatchord I'm really surprised that crossfading without phase adjustment works so well. I guess my attempt at phase matching is too complicated. 

Speed improvements with batch synthesis are really great. I prefer WaveRNN+batch generation to the inverse autoregressive flow approach used for wavenet, when it comes to fast synthesis.

**Comment by G-Wang** [comment]
@fatchord @geneing absolutely, this model in my opinion is miles better than the recent flow-based models, both in terms of model footprint, ease and speed of training, and generally good quality audio. And now we have batched speedups.

The next thing I'm looking into is pruning the model with something like https://github.com/NervanaSystems/distiller in Pytorch. And then we can potentially port everything over to CPU for faster than real time synthesis.

**Comment by geneing** [comment]
@fatchord mulaw sample sounds very good. There are barely noticeable artifacts in batched synthesis. 
Was the training with mulaw output slower than with raw?

**Comment by fatchord** [comment]
> Was the training with mulaw output slower than with raw?

@geneing What I did was I took a linear 9bit model (from notebook 4b) that was already trained to 500k steps and then just changed it to mu-law (fixing the offset mentioned above) and finetuned it to 500k steps along with a gradual reduction in lr (halved it every 100 - 150k steps). 

So it's kinda tricky to work backwards and guess how much training it really needs. I guess the only way to really know is to retrain from scratch and figure out a sensible lr schedule to use at the end of training. 

However I'm going to try and train a shallower and wider model this week because I want to start pruning weights. I think having two rnns and four fc layers, even with pruning, isn't a good idea for realtime+ cpu inference.

**Comment by G-Wang** [comment]
@fatchord what are you using for weight pruning? is it [distiller](https://github.com/NervanaSystems/distiller)? I'm curious for other methods if you're using something different.

**Comment by fatchord** [comment]
@G-Wang I'm coding up a Pruner() class that follows the pruning instructions in the paper. I'm not finished but you can have a look at how things are going in this notebook:

https://github.com/fatchord/WaveRNN/blob/master/Pruning%20-%20Scratchpad.ipynb

**Comment by azraelkuan** [comment]
@fatchord  for pruner, u can use the code https://github.com/mozilla/LPCNet/blob/b2606c7f7a8bcab22b2cb6d344285a14181e4367/src/lpcnet.py#L52-L87

i have adapted this code to my repo, the sparse tensor works well

and for the input, i suggest use the one hot coding for audio, i can get good result just using one gru and 2 upsample conv for 8bit

**Comment by syxu828** [comment]
Hi, thanks for sharing your excellent work! I just wonder what is the motivation of the model design: splitting the ResNet output into four parts, and combining them later.

**Comment by fatchord** [comment]
@azraelkuan Thanks for link. That LPCNet is a great paper/repo too!

@syxu828 The thinking behind that is to let the different layers all get a peek at the 'bigger picture' of where the signal was x frames ago and will be x frames in the future.

**Comment by hdmjdp** [comment]
@azraelkuan one hot coding for audio? such as wav[0] = 127, so it will be [0,0, 0,...,1,...0,0,0]? if this wav sample wll be a matrix instead of a vector?

**Comment by 1105060120** [comment]
What is the speed of synthesis when you use this model of WaveRNN?

**Comment by 1105060120** [comment]
@fatchord  how to export wavernn to C++ for real time synthesis.thank you

**Comment by 1105060120** [comment]
@G-Wang  how to export wavernn to C++,do you know how to do that?thank you
